### PR TITLE
More differentiated transformation #45

### PR DESCRIPTION
**Pull request by jerger**
This PR is just to keep you informed.

State of PR is:
* I've finished the rough part of refactoring work, 
* fine grained decoupling & realizing the new transformation config layout is open

### CODE REVIEW

**Comment by yogthos** [comment]
Thanks, so far this looks reasonable to me. Also, would you like me to add you as a maintainer on the repo, it looks like you've been doing more with it lately than myself. :)

**Comment by jerger** [comment]
Would be a honor to me ... but I hope you will stay in the loop. 
Beside of honor the real joy is the discussion with you, to get your ideas, hints & questions :-).

**Comment by yogthos** [comment]
Sounds like a plan, and yeah I'll be around to help out. Glad to hear I've been helpful. :)

File: src/main/mastodon_bot/transform.cljs [hunk 1]
```diff
@@ -73,12 +73,17 @@
...
    last-post-time any?]
-  (doseq [account accounts]
-    (twitter/user-timeline
-     twitter-config
-     account
-     (post-tweets mastodon-config last-post-time))))
+  (let [{:keys [source target]} transformation
+        {:keys [accounts include-rts? include-replies?]} source]
+    (doseq [account accounts]
+      (println account)
```
Comment (yogthos) [inline src/main/mastodon_bot/transform.cljs]: left over println?

**Comment by yogthos** [comment]
Looks good overall, I wonder if it might be better to make a general transform multimethod, and then implement it for each service?

File: src/main/mastodon_bot/transform.cljs (hunk 1 above)
Comment (jerger) [reply src/main/mastodon_bot/transform.cljs]: jupp ... moving carefully trhough :-)

**Comment by jerger** [comment]
Will move in baby-steps toward such a scenario.
But I'm unsure, whether multimethod will be the best solution (will have different signatures ...)

So let's get an impression & discussion in a more final state?

**Comment by yogthos** [comment]
Yeah sounds good, if the APIs aren't uniform might not make sense. There's also no requirement for it to be extensible.

**Comment by jerger** [comment]
But in meantime: "using spec to decouple functionalities & building responsibility areas is a really useful tool.

**Comment by jerger** [comment]
Now the transformation twitter-> mastodon is roughly at an end.
What do you think at this point ?

Next steps will be:
* introduce transformations for thumblr & rss
* decouple thumblr & rss-api from transformations
* improve testability
* inspect new transformation, whether we can find a good abstraction & pattern

**Comment by yogthos** [comment]
Looks good to me so far, and we should try to keep as much o the transformation logic generic as possible. At the end of the day we will end up with some text and media we extract from the source, and then we'll want to filter and transform that based on the content and metadata. So, we should do some massaging at each source to get the data into a common format, and then transform it after.

**Comment by jerger** [comment]
Do you have a good howto for converting the callback pattern to some kind of threading ?

```
(fkt params callback)
```

to

```
(-> (fkt params)
     callback)
```

**Comment by yogthos** [comment]
The callback would have to be called from within the function wouldn't it? [This](https://github.com/filipesilva/async-interop) got merged into the latest core.async though if you're trying to convert promises to sync style code. You can use `<p!` macro within the `go` blocks.

**Comment by jerger** [comment]
Cool :-). Do you've any concerns when I go more in direction of sync style? Just to get rid of my assync brain snarl :-)

Assync is imho fine in case of UI (but even there I've problems with  deep nested assyncs).  
And as we're on cli, maybe sync style is not the worst?

**Comment by yogthos** [comment]
Going towards sync style would definitely be preferable. Originally there wasn't a good way to deal with promises in cljs aside from async, but now that core.async has the `<p!` macro I'm very much in favor of using it.

**Comment by yogthos** [comment]
the update looks good, I like the approach of differentiating by `:source-type`

**Comment by jerger** [comment]
All apis are now driven by transform - rss & twitter is working for me :-)
At tumblr I've no account so I can not test so I'm not sure, whether refactoring is free of bugs.

At the definition of source-type I'm quite unhappy. As the following config shows, "source" & target are duplicated and very verbose.
```
:transform [{:source {:source-type :twitter ...}
                    :target {:target-type :mastodon ...}
```

On the other hand `(s/def ::source-type #{:twitter :rss :tumblr})` is nice & expressive. Do you know a way to get a shorter type description by not loosing expressive source / target types?

**Comment by yogthos** [comment]
Looks good, I'll take a look at trying out tumblr at some point since I've got an account I can use. And can't think of a better way to do the type description either. I think it's reasonable enough though.

**Comment by jerger** [comment]
:-)

If tumblr is working fine to you, the PR is ready to be merge from my point of view.

**Comment by yogthos** [comment]
Are the docs changes up to date as well?

**Comment by jerger** [comment]
jupp, readme is up to date exept `replacement`. There was no doc about it. Maybe you've an good example for?

**Comment by jerger** [comment]
reviewed doc and found the test.edn - I will refactor it to new layout ...

**Comment by yogthos** [comment]
I guess an example could be replacing could be something like removing `/amp` from a URL. And might be worth adding a paragraph on how transform works in the config. :)

**Comment by yogthos** [comment]
oh and looks like the start script in `package.json` needs to be updated to `"start": "./src/main/mastodon_bot/core.cljs"`, as well as `"bin": {"mastodon-bot": "..."}`.

**Comment by yogthos** [comment]
Hmm that actually doesn't seem to be working for me either, how are you launching the script on your end?

**Comment by yogthos** [comment]
It looks like we might actually have a problem with the use of orchestra, since it's not published via NPM. The library would have to be published using [these steps](https://andrearichiardi.com/blog/posts/lumo-npm-dependencies.html) in order to be usable with Lumo via NPM.

I don't think it would be a good situation where people have to download additional libs and install them locally themselves to use the bot. Sorry, I didn't catch that part earlier.

**Comment by jerger** [comment]
1. Updated package.json & adjusted paths. How can I test, whether this works correct?
    Invited you on my forked repo, so you will be able to do experiments directly.
2. How I start & test?: As I'm not familiar with npm, I use the repl: `(core/-main)`. For development & testing I use `shadow-cljs watch test`.
3. Regarding orchestra: I like the defn-spec. But maybe you know better alternatives there might be other implementations as well?
   If not, the best way would be to do a PR to orchestra?

**Comment by jerger** [comment]
Is there sth like a uberjar in the js world?
Maybe we can include the orchestra dep in our artefact?

**Comment by yogthos** [comment]
Unfortunately, there isn't any uberjar concept in js land. We could fork orchestra for now, and deploy it to npm so that we can use it as a dependency. It might be worth opening an issue with the project. One change that would need to happen is to put all the sources together in a `src` folder though. Perhaps that could be part of npm deploy script if the author wants to keep the project structure.

And for testing to see whether it works standalone you have to run the app via npm. The `start` script in package.json is what gets triggered when you do `npm start`. That's the standard way people would launch things in js world. We'd probably have to add another namespace that runs `-main` since Lumo doesn't automatically run it the way shadow-cljs does.

Try getting Lumo set up with npm locally just to see that you can run it that way.

**Comment by jerger** [comment]
* reg. uberjar: Do you know externs (https://shadow-cljs.github.io/docs/UsersGuide.html#externs)? Slightly sounds to be some kind of repacking?
* reg. orchestra fork: yes, would be the way & will do it (see https://github.com/jeaye/orchestra/issues/49 & https://github.com/DomainDrivenArchitecture/orchestra).
* reg lumo: I assume, lumo won't work without orchestra repackaged? But I will start to read about lumo ... :-)

**Comment by yogthos** [comment]
I haven't played with the externs packaging myself, so that'd be something to research. It might be an option to compile all of cljs code into js, and then publish the final artifact to NPM. Then people could just install the bot via NPM. If we did that then we don't have to worry about forking orchestra, since cljs would only be used with shadow-cljs at dev time.

**Comment by jerger** [comment]
Kind of js only uberjar - if it is working as we assume :-)
So seems to be the first option we can try ...

**Comment by yogthos** [comment]
Come to think of it, we don't even need to worry about externs, we just have to compile cljs using `:simple` optimizations flag. That will generate a single Js file that we can publish to NPM. It's not quite like uberjar, as it only contains cljs code from the project and the libraries we use, but won't contain NPM modules. But I think that's close enough in practical terms.

**Comment by yogthos** [comment]
here's the config I used to package with shadow

```clojure
{:source-paths ["src/main" "src/test"]
 :dependencies [[orchestra "2018.12.06-2"]]
 :builds       {:app {:target    :node-script
                    :output-to "target/mastodon-bot.js"
                    :main      mastodon-bot.core/main
                    :compiler-options {:optimizations :simple}}}}
```

**Comment by jerger** [comment]
cool, will test it :-)

**Comment by jerger** [comment]
Had an view on it. What doesn't work is:

`npx shadow-cljs release app`

It complains about missing (defn main). I think that's sth. lumo creates on the fly?
So if we leave the lumo integration out, then a 

```
shadow-cljs release app
node target/mastodon-bot.js 
```

is working on my machine (I know, famos last words :-)

If that's the way to go, I will add a main-fkt respecting the config-file parameter.

**Comment by yogthos** [comment]
Yeah, I think we can move off Lumo and just package the app using shadow-cljs instead then. And then we can take the artifact from `target/mastodon-bot.js` and publish it together with the `package.json` to NPM using this process https://andrearichiardi.com/blog/posts/lumo-npm-dependencies.html So, then people can just install the bot via NPM, and use `npm start` to run it.

**Comment by jerger** [comment]
First try to implement a simple spec-based cli - did some experiments with conform :-).
Works on my machine with `node target/mastodon-bot.js [-h] /path/to/config.edn`.

**Comment by yogthos** [comment]
Looking good. I ran `shadow-cljs release app` locally, and then `node target/mastodon-bot.js config.edn` looks to be running as expected. It'd probably be better to format errors in a more user friendly way though. :)

For example, probably would be better to say `unrecognized argument -a` and list the valid args instead of:

```
("-a") - failed: Extra input in: [0] spec: :mastodon-bot.core/args

Bad commandline arguments
```

similarly, ideally we should print a better message when config keys are wrong:

```
mastodon-bot/node_modules/mastodon-api/lib/mastodon.js:345
                    throw new Error('Mastodon config must include \'' + reqKey + '\' when using \'user_auth\'');
                    ^

Error: Mastodon config must include 'access_token' when using 'user_auth'
    at mastodon-bot/node_modules/mastodon-api/lib/mastodon.js:345:27
    at Array.forEach (<anonymous>)
    at Function._validateConfigOrThrow (mastodon-bot/node_modules/mastodon-api/lib/mastodon.js:343:46)
    at new Mastodon (mastodon-bot/node_modules/mastodon-api/lib/mastodon.js:41:18)
    at Object.mastodon_bot.mastodon_api.mastodon_client (mastodon-bot/target/mastodon-bot.js:4349:126)
    at Object.mastodon_bot.mastodon_api.get_mastodon_timeline (mastodon-bot/target/mastodon-bot.js:4391:96)
    at Object.mastodon_bot.core.transform_BANG_ (mastodon-bot/target/mastodon-bot.js:4692:157)
    at Function.mastodon_bot.core.main.cljs$core$IFn$_invoke$arity$variadic (mastodon-bot/target/mastodon-bot.js:4701:23)
    at Function.mastodon_bot.core.main.cljs$lang$applyTo (mastodon-bot/target/mastodon-bot.js:4701:157)
    at Function.cljs.core.apply.cljs$core$IFn$_invoke$arity$2 (mastodon-bot/target/mastodon-bot.js:1072:190)
```

Otherwise, I think it's mostly ready to go. Next step would be to add a script to push the js file in target to NPM.

**Comment by jerger** [comment]
reg . error messages I agree. Do you have an idea how to solve this sustainable?

**Comment by yogthos** [comment]
We could take a look at using expound, I've used it before and it tends to produce nice results from spec errors. I haven't tried using it with orchestra, but if it delegates to spec to print, then should work. https://github.com/bhb/expound

**Comment by jerger** [comment]
* expound is a really cool tip ... works like a charm :-)
* while improving error messages I found out that instrument seems not working as intended ... in some point in time I will do some investigation here / improve & PR orchestras instrumentation.
* having a working spec validation showed a wrong spec definition issue to me - thanx for this good hint for not being too lazy on errors :-)

**Comment by yogthos** [comment]
very nice, and glad a fix fell out of it ;)

**Comment by jerger** [comment]
:) so from my point of view, PR is ready for merge. What do you think?

**Comment by yogthos** [comment]
Yeah, I think so, do you want to try taking a look at publishing the compiled version to NPM. Then we could update the instructions in the readme to just be `npm install mastodon-bot` without needing to install Lumo.

**Comment by jerger** [comment]
How works rights management on npm? Is there sth like user(group)/project like it's on clojars?

Publishing will be a shell script containing some npm commands?

Should we add a travis build-step?

I can prepare this on my fork and my npm account and you will do the final publishing on your "official" npm ?

**Comment by yogthos** [comment]
I don't think npm has a concept of groups, so it would just be mastodon-bot. Here's a package.json file I used to publish markdown-clj

```javascript
{
  "name": "markdown-clj",
  "version": "1.10.1",
  "description": "Markdown parser written in Clojure/Script",
  "author": "Dmitri Sotnikov",
  "homepage": "https://github.com/yogthos/markdown-clj/",
  "license": "EPL-1.0",
  "directories": {
    "lib": "src"
  },
  "files": [
    "src/*"
  ],
  "keywords": [
    "cljs",
    "cljc",
    "markdown"
  ]
}
```
and it gets published to https://www.npmjs.com/package/markdown-clj and whoever pushes first will own the npm project, but it's possible to invite a maintainer. I use the same id `yogthos` on npm, so if you push from yours just invite me.

I haven't looked too closely at credentials management with npm, I've just been pushing from my local machine and it just asks for credentials on the prompt. On the web UI there is an auth tokens page, so it sounds like that's an option. If we can use travis, that would be pretty nifty. ;)

And yeah, I think we could make a shell script that would run shadow-cljs release, and then the npm commands to make a tarball and push it to npm. We might even be able to use the existing package.json for it.

**Comment by jerger** [comment]
okay, will do. But will also need some time to take the npm roadblock ... uff :-)

Btw. if you are interessed in, at https://github.com/DomainDrivenArchitecture/dda-masto-embed/blob/79f0a0e1cefdaff64eeb5b8c9f7f025d0c41f794/src/main/dda/masto_embed/app.cljs#L48 we are working on a more "clojure.async" - ish version of the mastodon api. Maybe we will find a way to share the synergy btwn. projects in some point in time - comments are welcome :-)

**Comment by yogthos** [comment]
oh interesting, speaking of, now that we can just use cljs libraries it's probably worth using core.async [<p!](https://clojurescript.org/guides/promise-interop#using-promises-with-core-async) macro. Would be a lot easier to maintain than all the callbacks. :)

And I'm definitely open to collaborating with dda-masto-embed, if we have some common functionality might as well reuse stuff. And I see that project is already using `<p!` as well. Looks like the way to go for sure then.

**Comment by jerger** [comment]
reg. collaboration: cool :-)

reg. core.async: I found it very difficult to work through promises, callbacks and channels. There is a lot to consider while using this stuff. But it's definitely an improvement compared to callback-hell.

I like the the idea of channels very - but compared to jvm async - to be forced to have an outer-callback fn like print or set-html! feels strange. 

And I've no feeling about nested go-blocks. So I will play around with ...

**Comment by yogthos** [comment]
I find the way I've been using it is making go blocks, and just using a let with the results of wrapping promises with <p! and haven't bothered with channels at all. But yeah whatever works best in your opinion, I'm pretty flexible on that front. :)

**Comment by jerger** [comment]
did a first try (you can find my build commands in .travis). But if I do `npm publish ./target/npm-build.tar --access public --tag dev0` I get a 

```
npm ERR! code E403
npm ERR! 403 403 Forbidden - PUT https://registry.npmjs.org/mastodon-bot - You do not have permission to publish "mastodon-bot". Are you logged in as the correct user?
```

npm login was successfull. Any idea?

**Comment by jerger** [comment]
Found the reason :-)
raboof did allready publish the mastodon-bot

@raboof are you willing & able to give the right to npm publish to yogthos (account on npm: yogthos) and me (account on npm: domain-driven-architecture) ?

**Comment by raboof** [comment]
> are you willing & able to give the right to npm publish to yogthos (account on npm: yogthos) and me (account on npm: domain-driven-architecture) ?

Sure. Let's create a `RELEASING.md` where we document how to release, so it's consistent/reproducible.

**Comment by jerger** [comment]
sth like this ?

I think it would be a good idea to automate many of these steps in future (either by travis or github actions). 
But let's start by a thumb manual process first.

**Comment by raboof** [comment]
> sth like this ?

Yeah, seems like a good start. Perhaps clarify that release builds should always be done from master?

> I think it would be a good idea to automate many of these steps in future (either by travis or github actions). But let's start by a thumb manual process first.

Agreed on both counts

**Comment by jerger** [comment]
Now I get

```
npm notice 
npm notice 📦  mastodon-bot@0.1.0-dev-0
npm notice === Tarball Contents === 
npm notice 0     ./               
npm notice 798B  ./package.json   
npm notice 2.3MB ./mastodon-bot.js
npm notice 5.3kB ./README.md      
npm notice === Tarball Details === 
npm notice name:          mastodon-bot                            
npm notice version:       0.1.0-dev-0                             
npm notice package size:  2.3 MB                                  
npm notice unpacked size: 2.3 MB                                  
npm notice shasum:        47d321d3ecb895cf1f6f8aea49a203d391ab3a42
npm notice integrity:     sha512-qCwSL4EQFC/Bf[...]P0LV2BZDfuIug==
npm notice total files:   4                                       
npm notice 
npm ERR! code E415
npm ERR! 415 Unsupported Media Type - PUT https://registry.npmjs.org/mastodon-bot - incorrect header check
```

Any idea?

**Comment by yogthos** [comment]
I haven't seen that before, maybe it doesn't like the version of zlib on your machine? https://stackoverflow.com/questions/45641629/node-zlib-incorrect-header-check?rq=1

**Comment by jerger** [comment]
cool, was the right idea :-)

I tried to publish a tar file, but npm wants a tar.gz file ... now publishing to npm works as described in release.md

**Comment by jerger** [comment]
added also some automation - atmo the actions are running on my forked repo (https://github.com/DomainDrivenArchitecture/mastodon-bot/runs/814067106?check_suite_focus=true). 

As I'm not sure, how github handles fork / upstream activities now is the good moment to merge the PR and continue the automation work in master?

I think it should be quite easy possible to upload the `mastodon-bot.js` to github-releases section. As a second step we can take care about getting a start able npm upload.

**Comment by yogthos** [comment]
Automation is a nice touch, and yeah let's merge this all in. I think we're in a good spot now. Nice work on getting this all set up. :)

**Comment by yogthos** [comment]
And yeah let's publish the js artifact to releases and we can add a link in the readme for the download until we get npm figured out.

**Comment by yogthos** [comment]
One big advantage here will be having stable release artifacts going forward as well regardless of what we're doing in the repo.

**Pull request merged**